File: Amazons_simple_init.cpp
#include<random>
#include <iostream>
#include <string>
#include <cstdlib>
#include <ctime>
#include <vector>
#include<queue>
#include<algorithm>
#include<cmath>
#include <map>
#include<cstring>

#define GRIDSIZE 8
#define OBSTACLE 2
#define judge_black 0
#define judge_white 1
#define grid_black 1
#define grid_white -1

clock_t start_time;
bool stop_searching = false;

// --- 计时函数 ---
void startTimer() {
    start_time = clock();
}

bool timeIsUp() {
    const double elapsed = (double)(clock() - start_time) / CLOCKS_PER_SEC;
    return elapsed > 0.93;
}

// 历史启发表
int history[2][GRIDSIZE][GRIDSIZE] = {0};

const int MAX_DEPTH = 1000;

using namespace std;

int currBotColor;
int gridInfo[GRIDSIZE][GRIDSIZE] = { 0 };
int dx[] = { -1,-1,-1,0,0,1,1,1 };
int dy[] = { -1,0,1,-1,1,-1,0,1 };

struct Point {
    int x;
    int y;
    Point() : x(0), y(0) {}
    Point(int x0, int y0) : x(x0), y(y0) {}
};

struct Move {
    Point initgrid;
    Point newgrid;
    Point arows;
    Move(Point init, Point newgrid, Point arows) : initgrid(init), newgrid(newgrid), arows(arows) {}
    Move() {}
};

// 置换表
struct TTEntry {
    unsigned long long key;
    int depth;
    double value;
    int flag;
    Move bestMove;
};

const int TT_SIZE = 2 << 20;
TTEntry TT[TT_SIZE];

unsigned long long zobristTable[GRIDSIZE][GRIDSIZE][4];
unsigned long long currentHash = 0;

void initZobrist() {
    std::mt19937_64 rng(12345);
    for (int i = 0; i < GRIDSIZE; i++) {
        for (int j = 0; j < GRIDSIZE; j++) {
            for (int k = 0; k < 4; k++) {
                zobristTable[i][j][k] = rng();
            }
        }
    }
}

int getPieceIndex(int val) {
    if (val == 0) return 0;
    if (val == grid_black) return 1;
    if (val == grid_white) return 2;
    if (val == OBSTACLE) return 3;
    return 0;
}

unsigned long long computeInitialHash() {
    unsigned long long h = 0;
    for (int i = 0; i < GRIDSIZE; i++) {
        for (int j = 0; j < GRIDSIZE; j++) {
            int pieceIdx = getPieceIndex(gridInfo[i][j]);
            if (pieceIdx != 0) {
                h ^= zobristTable[i][j][pieceIdx];
            }
        }
    }
    return h;
}

unsigned long long updateHash(unsigned long long oldHash, int x0, int y0, int x1, int y1, int x2, int y2, int color) {
    unsigned long long newHash = oldHash;
    newHash ^= zobristTable[x0][y0][getPieceIndex(color)];
    newHash ^= zobristTable[x1][y1][getPieceIndex(color)];
    newHash ^= zobristTable[x2][y2][getPieceIndex(OBSTACLE)];
    return newHash;
}

inline bool inMap(int x, int y) {
    if (x < 0 || x >= GRIDSIZE || y < 0 || y >= GRIDSIZE)
        return false;
    return true;
}

bool ProcStep(int x0, int y0, int x1, int y1, int x2, int y2, int color, bool check_only) {
    if ((!inMap(x0, y0)) || (!inMap(x1, y1)) || (!inMap(x2, y2)))
        return false;
    if (gridInfo[x0][y0] != color || gridInfo[x1][y1] != 0)
        return false;
    if ((gridInfo[x2][y2] != 0) && !(x2 == x0 && y2 == y0))
        return false;
    if (!check_only) {
        gridInfo[x0][y0] = 0;
        gridInfo[x1][y1] = color;
        gridInfo[x2][y2] = OBSTACLE;
    }
    return true;
}

bool LegalStep(int x, int y, int grid_info[GRIDSIZE][GRIDSIZE]) {
    if (!inMap(x, y)) return false;
    else if (grid_info[x][y] == OBSTACLE || grid_info[x][y] == grid_black || grid_info[x][y] == grid_white) {
        return false;
    }
    return true;
}

vector<Point> get_move_pos(Point cu_point, int tpgrid[GRIDSIZE][GRIDSIZE]) {
    vector<Point> vecpo;
    for (int d = 0; d < 8; d++) {
        int px = cu_point.x;
        int py = cu_point.y;
        while (true) {
            px += dx[d];
            py += dy[d];
            if (LegalStep(px, py, tpgrid)) {
                vecpo.emplace_back(px, py);
            }
            else {
                break;
            }
        }
    }
    return vecpo;
}

vector<Point> get_arrow_pos(int temp_grid[GRIDSIZE][GRIDSIZE], Point newpoint) {
    vector<Point> vecpo;
    for (int d = 0; d < 8; d++) {
        int px = newpoint.x;
        int py = newpoint.y;
        while (true) {
            px += dx[d];
            py += dy[d];
            if (LegalStep(px, py, temp_grid)) {
                vecpo.emplace_back(px, py);
            }
            else {
                break;
            }
        }
    }
    return vecpo;
}

vector<Move> get_valid_moves(int color, int tpgrid[GRIDSIZE][GRIDSIZE]) {
    vector<Move> all_valid_moves;
    vector<Point> starts;

    for (int x = 0; x < GRIDSIZE; x++) {
        for (int y = 0; y < GRIDSIZE; y++) {
            if (tpgrid[x][y] == color) {
                starts.emplace_back(x, y);
            }
        }
    }

    for (auto& p : starts) {
        vector<Point> move_positions = get_move_pos(p, tpgrid);

        for (auto& new_p : move_positions) {
            int temp_grid[GRIDSIZE][GRIDSIZE];
            memcpy(temp_grid, tpgrid, sizeof(int[GRIDSIZE][GRIDSIZE]));
            temp_grid[p.x][p.y] = 0;
            temp_grid[new_p.x][new_p.y] = color;

            vector<Point> arrows_position = get_arrow_pos(temp_grid, new_p);

            for (auto& arrow : arrows_position) {
                if (arrow.x == new_p.x && arrow.y == new_p.y) continue;
                all_valid_moves.emplace_back(p, new_p, arrow);
            }
        }
    }

    return all_valid_moves;
}

inline int centerScore(const Point& p) {
    return 7 - (abs(p.x - 3) + abs(p.y - 3));
}

int quickMoveScore(const Move& m, int color, int tpgrid[GRIDSIZE][GRIDSIZE]) {
    int score = 0;
    score += centerScore(m.newgrid) * 6;
    score += centerScore(m.arows) * 2;
    score += max(abs(m.newgrid.x - m.initgrid.x), abs(m.newgrid.y - m.initgrid.y)) * 2;

    int opp = -color;
    for (int d = 0; d < 8; ++d) {
        int nx = m.arows.x + dx[d], ny = m.arows.y + dy[d];
        if (inMap(nx, ny) && tpgrid[nx][ny] == opp) score += 8;
    }
    return score;
}

void computeDistances(int color, int tpgrid[GRIDSIZE][GRIDSIZE], int distMap[GRIDSIZE][GRIDSIZE], int mode) {
    for (int i = 0; i < GRIDSIZE; ++i)
        for (int j = 0; j < GRIDSIZE; ++j)
            distMap[i][j] = 100;

    queue<pair<int, int>> q;
    for (int i = 0; i < GRIDSIZE; ++i) {
        for (int j = 0; j < GRIDSIZE; ++j) {
            if (tpgrid[i][j] == color) {
                distMap[i][j] = 0;
                q.push({ i, j });
            }
        }
    }

    while (!q.empty()) {
        pair<int, int> curr = q.front();
        q.pop();
        int nextDist = distMap[curr.first][curr.second] + 1;
        for (int d = 0; d < 8; d++) {
            int nx = curr.first + dx[d];
            int ny = curr.second + dy[d];

            if (mode == 1) {
                while (inMap(nx, ny) && tpgrid[nx][ny] == 0) {
                    if (distMap[nx][ny] > nextDist) {
                        distMap[nx][ny] = nextDist;
                        q.push({ nx, ny });
                    }
                    else if (distMap[nx][ny] < nextDist) {
                        break;
                    }
                    nx += dx[d];
                    ny += dy[d];
                }
            }
            else {
                if (inMap(nx, ny) && tpgrid[nx][ny] == 0) {
                    if (distMap[nx][ny] > distMap[curr.first][curr.second] + 1) {
                        distMap[nx][ny] = distMap[curr.first][curr.second] + 1;
                        q.push({ nx, ny });
                    }
                }
            }
        }
    }
}

int reachableCountFromQueen(const Point& q, int tpgrid[GRIDSIZE][GRIDSIZE]) {
    int count = 0;
    for (int d = 0; d < 8; ++d) {
        int x = q.x + dx[d], y = q.y + dy[d];
        while (inMap(x, y) && tpgrid[x][y] == 0) {
            ++count;
            x += dx[d];
            y += dy[d];
        }
    }
    return count;
}

double getMobilityScore(int color, int tpgrid[GRIDSIZE][GRIDSIZE]) {
    int totalReach = 0;
    int minReach = 9999;
    int queenCnt = 0;

    for (int x = 0; x < GRIDSIZE; ++x) {
        for (int y = 0; y < GRIDSIZE; ++y) {
            if (tpgrid[x][y] != color) continue;
            ++queenCnt;
            int reach = reachableCountFromQueen(Point(x, y), tpgrid);
            totalReach += reach;
            minReach = min(minReach, reach);
        }
    }

    if (queenCnt == 0 || minReach == 9999) minReach = 0;
    return (double)totalReach + (double)minReach;
}

double evaluate(int tpgrid[GRIDSIZE][GRIDSIZE], int turnID) {
    int whiteDistQ[GRIDSIZE][GRIDSIZE], blackDistQ[GRIDSIZE][GRIDSIZE];
    int whiteDistK[GRIDSIZE][GRIDSIZE], blackDistK[GRIDSIZE][GRIDSIZE];

    computeDistances(grid_white, tpgrid, whiteDistQ, 1);
    computeDistances(grid_black, tpgrid, blackDistQ, 1);
    computeDistances(grid_white, tpgrid, whiteDistK, 2);
    computeDistances(grid_black, tpgrid, blackDistK, 2);

    double t1 = 0, t2 = 0, p1 = 0, p2 = 0;

    for (int i = 0; i < GRIDSIZE; i++) {
        for (int j = 0; j < GRIDSIZE; j++) {
            if (tpgrid[i][j] != 0) continue;

            if (whiteDistQ[i][j] < blackDistQ[i][j]) t1 += 1.0;
            else if (whiteDistQ[i][j] > blackDistQ[i][j]) t1 -= 1.0;

            if (whiteDistK[i][j] < blackDistK[i][j]) t2 += 1.0;
            else if (whiteDistK[i][j] > blackDistK[i][j]) t2 -= 1.0;

            p1 += (pow(2, -whiteDistQ[i][j]) - pow(2, -blackDistQ[i][j]));
            double diffK = (double)(blackDistK[i][j] - whiteDistK[i][j]) / 6.0;
            p2 += max(-1.0, min(1.0, diffK));
        }
    }

    double m = getMobilityScore(grid_white, tpgrid) - getMobilityScore(grid_black, tpgrid);

    double a, b, c, d, e;
    if (turnID <= 20) {
        a = 0.14; b = 0.37; c = 0.13; d = 0.13; e = 0.20;
    }
    else if (turnID <= 49) {
        a = 0.25; b = 0.30; c = 0.20; d = 0.20; e = 0.05;
    }
    else {
        a = 0.80; b = 0.10; c = 0.05; d = 0.05; e = 0.00;
    }

    double score = a * t1 + b * t2 + c * p1 + d * p2 + e * m;
    return (currBotColor == grid_white) ? score : -score;
}

void get_newgrid(int tmpgrid[GRIDSIZE][GRIDSIZE], Move& move, int color) {
    Point start = move.initgrid, newg = move.newgrid, arrow = move.arows;
    tmpgrid[start.x][start.y] = 0;
    tmpgrid[newg.x][newg.y] = color;
    tmpgrid[arrow.x][arrow.y] = 2;
}

bool isSameMove(const Move& m1, const Move& m2) {
    return m1.initgrid.x == m2.initgrid.x && m1.initgrid.y == m2.initgrid.y &&
        m1.newgrid.x == m2.newgrid.x && m1.newgrid.y == m2.newgrid.y &&
        m1.arows.x == m2.arows.x && m1.arows.y == m2.arows.y;
}

double terminalScore(bool isMaxPlayer) {
    return isMaxPlayer ? -1e8 : 1e8;
}

// ========================== 最终 MinMax（带历史启发 + TT + αβ）==========================
double MinMax(int grid[GRIDSIZE][GRIDSIZE], int depth, bool isMax, int turnID, double arfa, double beta, unsigned long long h) {
    if (stop_searching || timeIsUp()) {
        stop_searching = true;
        return evaluate(grid, turnID);
    }

    int index = h & (TT_SIZE - 1);
    Move bestMoveFromTT;
    bool hasBestMoveFromTT = false;
    const double alphaOrig = arfa;
    const double betaOrig = beta;

    if (TT[index].key == h) {
        bestMoveFromTT = TT[index].bestMove;
        hasBestMoveFromTT = true;

        if (TT[index].depth >= depth) {
            if (TT[index].flag == 0) return TT[index].value;
            if (TT[index].flag == 1 && TT[index].value >= beta) return TT[index].value;
            if (TT[index].flag == 2 && TT[index].value <= arfa) return TT[index].value;
        }
    }

    if (depth == 0) {
        return evaluate(grid, turnID);
    }

    Move bestMoveInThisNode;
    bool foundBest = false;

    if (isMax) {
        vector<Move> moves = get_valid_moves(currBotColor, grid);
        if (moves.empty()) return terminalScore(true);
        double currentmax = -1e9;

        stable_sort(moves.begin(), moves.end(), [&](const Move& a, const Move& b) {
            int cidx = (currBotColor == grid_black) ? 0 : 1;
            int ha = history[cidx][a.initgrid.x][a.initgrid.y];
            int hb = history[cidx][b.initgrid.x][b.initgrid.y];
            if (ha != hb) return ha > hb;
            return quickMoveScore(a, currBotColor, grid) > quickMoveScore(b, currBotColor, grid);
        });

        if (hasBestMoveFromTT) {
            for (size_t i = 0; i < moves.size(); i++) {
                if (isSameMove(moves[i], bestMoveFromTT)) {
                    swap(moves[0], moves[i]);
                    break;
                }
            }
        }

        for (auto& m : moves) {
            int tmpgrid[GRIDSIZE][GRIDSIZE];
            memcpy(tmpgrid, grid, sizeof(int[GRIDSIZE][GRIDSIZE]));
            get_newgrid(tmpgrid, m, currBotColor);

            unsigned long long next_h = updateHash(h, m.initgrid.x, m.initgrid.y,
                m.newgrid.x, m.newgrid.y,
                m.arows.x, m.arows.y, currBotColor);

            double tpscore = MinMax(tmpgrid, depth - 1, false, turnID, arfa, beta, next_h);

            if (tpscore > currentmax) {
                currentmax = tpscore;
                bestMoveInThisNode = m;
                foundBest = true;
            }

            if (currentmax > beta) {
                int cidx = (currBotColor == grid_black) ? 0 : 1;
                history[cidx][m.initgrid.x][m.initgrid.y] += 120;
                break;
            }
            if (currentmax > arfa) arfa = currentmax;
        }

        if (TT[index].key != h || depth >= TT[index].depth) {
            TT[index].key = h;
            TT[index].depth = depth;
            TT[index].value = currentmax;
            if (foundBest) TT[index].bestMove = bestMoveInThisNode;
            if (currentmax <= alphaOrig)      TT[index].flag = 2;
            else if (currentmax >= betaOrig)  TT[index].flag = 1;
            else                              TT[index].flag = 0;
        }
        return currentmax;
    }
    else {
        vector<Move> moves = get_valid_moves((-1) * currBotColor, grid);
        if (moves.empty()) return terminalScore(false);
        double currentmin = 1e9;

        stable_sort(moves.begin(), moves.end(), [&](const Move& a, const Move& b) {
            int oppColor = -currBotColor;
            int cidx = (oppColor == grid_black) ? 0 : 1;
            int ha = history[cidx][a.initgrid.x][a.initgrid.y];
            int hb = history[cidx][b.initgrid.x][b.initgrid.y];
            if (ha != hb) return ha > hb;
            return quickMoveScore(a, oppColor, grid) > quickMoveScore(b, oppColor, grid);
        });

        if (hasBestMoveFromTT) {
            for (size_t i = 0; i < moves.size(); i++) {
                if (isSameMove(moves[i], bestMoveFromTT)) {
                    swap(moves[0], moves[i]);
                    break;
                }
            }
        }

        for (auto& m : moves) {
            int tmpgrid[GRIDSIZE][GRIDSIZE];
            memcpy(tmpgrid, grid, sizeof(int[GRIDSIZE][GRIDSIZE]));
            get_newgrid(tmpgrid, m, (-1) * currBotColor);

            unsigned long long next_h = updateHash(h, m.initgrid.x, m.initgrid.y,
                m.newgrid.x, m.newgrid.y,
                m.arows.x, m.arows.y, (-1) * currBotColor);

            double tpscore = MinMax(tmpgrid, depth - 1, true, turnID, arfa, beta, next_h);

            if (tpscore < currentmin) {
                currentmin = tpscore;
                bestMoveInThisNode = m;
                foundBest = true;
            }

            if (currentmin < arfa) {
                int oppColor = -currBotColor;
                int cidx = (oppColor == grid_black) ? 0 : 1;
                history[cidx][m.initgrid.x][m.initgrid.y] += 120;
                break;
            }
            if (currentmin < beta) beta = currentmin;
        }

        if (TT[index].key != h || depth >= TT[index].depth) {
            TT[index].key = h;
            TT[index].depth = depth;
            TT[index].value = currentmin;
            if (foundBest) TT[index].bestMove = bestMoveInThisNode;
            if (currentmin <= alphaOrig)      TT[index].flag = 2;
            else if (currentmin >= betaOrig)  TT[index].flag = 1;
            else                              TT[index].flag = 0;
        }
        return currentmin;
    }
}

double MTDF(unsigned long long h, double f, int depth, int turnID) {
    double g = f;
    double upperbound = 1e9;
    double lowerbound = -1e9;

    while (lowerbound < upperbound - 0.001) {
        double beta = (g == lowerbound) ? g + 0.01 : g;
        g = MinMax(gridInfo, depth, true, turnID, beta - 0.01, beta, h);
        if (stop_searching) break;
        if (g < beta) upperbound = g;
        else lowerbound = g;
    }
    return g;
}

int main() {
    int x0, y0, x1, y1, x2, y2;

    gridInfo[0][2] = grid_black;
    gridInfo[2][0] = grid_black;
    gridInfo[5][0] = grid_black;
    gridInfo[7][2] = grid_black;

    gridInfo[0][5] = grid_white;
    gridInfo[2][7] = grid_white;
    gridInfo[5][7] = grid_white;
    gridInfo[7][5] = grid_white;

    initZobrist();
    currentHash = computeInitialHash();

    int turnID;
    cin >> turnID;

    currBotColor = grid_white;
    for (int i = 0; i < turnID; i++) {
        cin >> x0 >> y0 >> x1 >> y1 >> x2 >> y2;
        if (x0 == -1)
            currBotColor = grid_black;
        else {
            ProcStep(x0, y0, x1, y1, x2, y2, -currBotColor, false);
            currentHash = updateHash(currentHash, x0, y0, x1, y1, x2, y2, -currBotColor);
        }

        if (i < turnID - 1) {
            cin >> x0 >> y0 >> x1 >> y1 >> x2 >> y2;
            if (x0 >= 0) {
                ProcStep(x0, y0, x1, y1, x2, y2, currBotColor, false);
                currentHash = updateHash(currentHash, x0, y0, x1, y1, x2, y2, currBotColor);
            }
        }
    }

    startTimer();
    stop_searching = false;
    vector<Move> rootMoves = get_valid_moves(currBotColor, gridInfo);
    Move overallBestMove = rootMoves.empty() ? Move() : rootMoves[0];
    double last_score = 0;

    for (int d = 1; d <= MAX_DEPTH; d++) {
        memset(history, 0, sizeof(history));
        double current_score = MTDF(currentHash, last_score, d, turnID);

        if (stop_searching) break;
        last_score = current_score;

        int index = currentHash & (TT_SIZE - 1);
        if (TT[index].key == currentHash) {
            overallBestMove = TT[index].bestMove;
        }
    }

    bool legalBest = false;
    for (auto& m : rootMoves) {
        if (isSameMove(m, overallBestMove)) {
            legalBest = true;
            break;
        }
    }
    if (!legalBest) overallBestMove = rootMoves[0];

    int startX = overallBestMove.initgrid.x;
    int startY = overallBestMove.initgrid.y;
    int resultX = overallBestMove.newgrid.x;
    int resultY = overallBestMove.newgrid.y;
    int obstacleX = overallBestMove.arows.x;
    int obstacleY = overallBestMove.arows.y;

    cout << startX << ' ' << startY << ' ' << resultX << ' ' << resultY << ' ' << obstacleX << ' ' << obstacleY << endl;
    return 0;
}

File: Amazons_simple_init.rs
use std::cmp::{max, min};
use std::collections::VecDeque;
use std::io::{self, Read};
use std::time::Instant;

const GRIDSIZE: usize = 8;
const OBSTACLE: i32 = 2;
const GRID_BLACK: i32 = 1;
const GRID_WHITE: i32 = -1;
const MAX_DEPTH: usize = 10_000_000;
const TT_SIZE: usize = 2 << 20;

static DX: [i32; 8] = [-1, -1, -1, 0, 0, 1, 1, 1];
static DY: [i32; 8] = [-1, 0, 1, -1, 1, -1, 0, 1];

#[derive(Clone, Copy, Default, PartialEq, Eq)]
struct Point { x: usize, y: usize }

#[derive(Clone, Copy, Default, PartialEq, Eq)]
struct Move { initgrid: Point, newgrid: Point, arows: Point }

#[derive(Clone, Copy, Default)]
struct TTEntry { key: u64, depth: usize, value: f64, flag: i32, best_move: Move }

struct Engine {
    start_time: Instant,
    stop_searching: bool,
    curr_bot_color: i32,
    grid_info: [[i32; GRIDSIZE]; GRIDSIZE],
    history_from: [[[i32; GRIDSIZE]; GRIDSIZE]; 2],
    history_to: [[[i32; GRIDSIZE]; GRIDSIZE]; 2],
    tt: Vec<TTEntry>,
    zobrist: [[[u64; 4]; GRIDSIZE]; GRIDSIZE],
    current_hash: u64,
}

impl Engine {
    fn new() -> Self {
        let mut s = Self {
            start_time: Instant::now(),
            stop_searching: false,
            curr_bot_color: GRID_WHITE,
            grid_info: [[0; GRIDSIZE]; GRIDSIZE],
            history_from: [[[0; GRIDSIZE]; GRIDSIZE]; 2],
            history_to: [[[0; GRIDSIZE]; GRIDSIZE]; 2],
            tt: vec![TTEntry::default(); TT_SIZE],
            zobrist: [[[0; 4]; GRIDSIZE]; GRIDSIZE],
            current_hash: 0,
        };
        s.init_board();
        s.init_zobrist();
        s.current_hash = s.compute_initial_hash();
        s
    }
    fn init_board(&mut self) {
        self.grid_info[0][2] = GRID_BLACK; self.grid_info[2][0] = GRID_BLACK;
        self.grid_info[5][0] = GRID_BLACK; self.grid_info[7][2] = GRID_BLACK;
        self.grid_info[0][5] = GRID_WHITE; self.grid_info[2][7] = GRID_WHITE;
        self.grid_info[5][7] = GRID_WHITE; self.grid_info[7][5] = GRID_WHITE;
    }
    fn time_is_up(&self) -> bool { self.start_time.elapsed().as_secs_f64() > 0.93 }
    fn in_map(x: i32, y: i32) -> bool { x >= 0 && x < GRIDSIZE as i32 && y >= 0 && y < GRIDSIZE as i32 }
    fn legal_step(x: i32, y: i32, grid: &[[i32; GRIDSIZE]; GRIDSIZE]) -> bool {
        if !Self::in_map(x,y) { return false; }
        grid[x as usize][y as usize] == 0
    }
    fn get_move_pos(cu: Point, grid: &[[i32; GRIDSIZE]; GRIDSIZE]) -> Vec<Point> {
        let mut v = vec![];
        for d in 0..8 {
            let (mut px, mut py) = (cu.x as i32, cu.y as i32);
            loop {
                px += DX[d]; py += DY[d];
                if Self::legal_step(px, py, grid) { v.push(Point{ x:px as usize, y:py as usize}); } else { break; }
            }
        }
        v
    }
    fn get_arrow_pos(grid: &[[i32; GRIDSIZE]; GRIDSIZE], p: Point) -> Vec<Point> { Self::get_move_pos(p, grid) }
    fn get_valid_moves(color: i32, grid: &[[i32; GRIDSIZE]; GRIDSIZE]) -> Vec<Move> {
        let mut starts = vec![];
        for x in 0..GRIDSIZE { for y in 0..GRIDSIZE { if grid[x][y] == color { starts.push(Point{x,y}); } } }
        let mut all = vec![];
        for p in starts {
            for new_p in Self::get_move_pos(p, grid) {
                let mut tmp = *grid;
                tmp[p.x][p.y] = 0; tmp[new_p.x][new_p.y] = color;
                for arrow in Self::get_arrow_pos(&tmp, new_p) {
                    if arrow != new_p { all.push(Move{initgrid:p,newgrid:new_p,arows:arrow}); }
                }
            }
        }
        all
    }
    fn center_score(p: Point) -> i32 { 7 - ((p.x as i32 - 3).abs() + (p.y as i32 - 3).abs()) }
    fn quick_move_score(m: Move, color: i32, grid: &[[i32; GRIDSIZE]; GRIDSIZE]) -> i32 {
        let mut score = 0;
        score += Self::center_score(m.newgrid) * 6 + Self::center_score(m.arows) * 2;
        score += max((m.newgrid.x as i32 - m.initgrid.x as i32).abs(), (m.newgrid.y as i32 - m.initgrid.y as i32).abs()) * 2;
        let opp = -color;
        for d in 0..8 {
            let nx = m.arows.x as i32 + DX[d]; let ny = m.arows.y as i32 + DY[d];
            if Self::in_map(nx, ny) && grid[nx as usize][ny as usize] == opp { score += 8; }
        }
        score
    }
    fn compute_distances(color:i32, grid:&[[i32;GRIDSIZE];GRIDSIZE], dist:&mut [[i32;GRIDSIZE];GRIDSIZE], mode:i32){
        for i in 0..GRIDSIZE { for j in 0..GRIDSIZE { dist[i][j]=100; } }
        let mut q=VecDeque::new();
        for i in 0..GRIDSIZE { for j in 0..GRIDSIZE { if grid[i][j]==color {dist[i][j]=0;q.push_back((i,j));}}}
        while let Some((x,y))=q.pop_front(){
            let next=dist[x][y]+1;
            for d in 0..8 {
                let mut nx=x as i32 + DX[d]; let mut ny=y as i32 + DY[d];
                if mode==1 { while Self::in_map(nx,ny) && grid[nx as usize][ny as usize]==0 { let (ux,uy)=(nx as usize,ny as usize); if dist[ux][uy]>next {dist[ux][uy]=next;q.push_back((ux,uy));} else if dist[ux][uy]<next {break;} nx+=DX[d]; ny+=DY[d]; } }
                else if Self::in_map(nx,ny) && grid[nx as usize][ny as usize]==0 { let (ux,uy)=(nx as usize,ny as usize); if dist[ux][uy]>dist[x][y]+1 {dist[ux][uy]=dist[x][y]+1;q.push_back((ux,uy));} }
            }
        }
    }
    fn evaluate(&self, grid:&[[i32;GRIDSIZE];GRIDSIZE], turn_id:i32)->f64{
        let (mut wq,mut bq,mut wk,mut bk)=([[0;GRIDSIZE];GRIDSIZE],[[0;GRIDSIZE];GRIDSIZE],[[0;GRIDSIZE];GRIDSIZE],[[0;GRIDSIZE];GRIDSIZE]);
        Self::compute_distances(GRID_WHITE,grid,&mut wq,1); Self::compute_distances(GRID_BLACK,grid,&mut bq,1);
        Self::compute_distances(GRID_WHITE,grid,&mut wk,2); Self::compute_distances(GRID_BLACK,grid,&mut bk,2);
        let (mut t1,mut t2,mut p1,mut p2)=(0.0,0.0,0.0,0.0);
        for i in 0..GRIDSIZE { for j in 0..GRIDSIZE { if grid[i][j]!=0 {continue;} if wq[i][j]<bq[i][j]{t1+=1.0}else if wq[i][j]>bq[i][j]{t1-=1.0}; if wk[i][j]<bk[i][j]{t2+=1.0}else if wk[i][j]>bk[i][j]{t2-=1.0}; p1 += 2f64.powi(-wq[i][j]) - 2f64.powi(-bq[i][j]); let diff=(bk[i][j]-wk[i][j]) as f64/6.0; p2 += diff.clamp(-1.0,1.0);} }
        let (a,b,c,d,e,f)= if turn_id<=20 {(0.14,0.33,0.13,0.17,0.17,0.10)} else if turn_id<=49 {(0.24,0.28,0.18,0.18,0.04,0.08)} else {(0.72,0.10,0.05,0.05,0.0,0.08)};
        let score=a*t1+b*t2+c*p1+d*p2+e*0.0+f*0.0;
        if self.curr_bot_color==GRID_WHITE {score} else {-score}
    }
    fn init_zobrist(&mut self){let mut x: u64 = 88172645463393265; for i in 0..GRIDSIZE {for j in 0..GRIDSIZE {for k in 0..4 {x ^= x<<7; x ^= x>>9; self.zobrist[i][j][k]=x;}}}}
    fn piece_idx(v:i32)->usize{if v==0 {0}else if v==GRID_BLACK {1}else if v==GRID_WHITE {2}else {3}}
    fn compute_initial_hash(&self)->u64{let mut h=0; for i in 0..GRIDSIZE {for j in 0..GRIDSIZE {let p=Self::piece_idx(self.grid_info[i][j]); if p!=0 {h ^= self.zobrist[i][j][p];}}} h}
    fn update_hash(&self, old:u64,m:Move,color:i32)->u64{let mut h=old; h ^= self.zobrist[m.initgrid.x][m.initgrid.y][Self::piece_idx(color)]; h ^= self.zobrist[m.newgrid.x][m.newgrid.y][Self::piece_idx(color)]; h ^= self.zobrist[m.arows.x][m.arows.y][Self::piece_idx(OBSTACLE)]; h}
}

fn main() {
    let mut engine = Engine::new();
    let mut input = String::new(); io::stdin().read_to_string(&mut input).unwrap();
    let mut it = input.split_whitespace();
    let turn_id: i32 = it.next().unwrap_or("0").parse().unwrap_or(0);
    for i in 0..turn_id {
        let vals: Vec<i32> = (0..6).map(|_| it.next().unwrap_or("-1").parse().unwrap_or(-1)).collect();
        if vals[0] == -1 { engine.curr_bot_color = GRID_BLACK; }
        else {
            let m = Move{initgrid:Point{x:vals[0] as usize,y:vals[1] as usize},newgrid:Point{x:vals[2] as usize,y:vals[3] as usize},arows:Point{x:vals[4] as usize,y:vals[5] as usize}};
            engine.grid_info[m.initgrid.x][m.initgrid.y]=0; engine.grid_info[m.newgrid.x][m.newgrid.y]=-engine.curr_bot_color; engine.grid_info[m.arows.x][m.arows.y]=OBSTACLE;
            engine.current_hash = engine.update_hash(engine.current_hash,m,-engine.curr_bot_color);
        }
        if i < turn_id-1 { let _ : Vec<i32> = (0..6).map(|_| it.next().unwrap_or("-1").parse().unwrap_or(-1)).collect(); }
    }
    let mut root = Engine::get_valid_moves(engine.curr_bot_color,&engine.grid_info);
    if root.is_empty() { println!("-1 -1 -1 -1 -1 -1"); return; }
    root.sort_by_key(|m| -Engine::quick_move_score(*m, engine.curr_bot_color, &engine.grid_info));
    let best = root[0];
    println!("{} {} {} {} {} {}", best.initgrid.x,best.initgrid.y,best.newgrid.x,best.newgrid.y,best.arows.x,best.arows.y);
}
